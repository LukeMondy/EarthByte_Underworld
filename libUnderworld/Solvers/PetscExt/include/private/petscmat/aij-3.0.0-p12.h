
#if !defined(__AIJ_H)
#define __AIJ_H

#include "private/matimpl.h"
/*  
    Struct header shared by SeqAIJ, SeqBAIJ and SeqSBAIJ matrix formats
*/
#define SEQAIJHEADER(datatype)	\
  PetscTruth        roworiented;      /* if true, row-oriented input, default */\
  PetscInt          nonew;            /* 1 don't add new nonzeros, -1 generate error on new */\
  PetscInt          nounused;         /* -1 generate error on unused space */\
  PetscTruth        singlemalloc;     /* if true a, i, and j have been obtained with one big malloc */\
  PetscInt          maxnz;            /* allocated nonzeros */\
  PetscInt          *imax;            /* maximum space allocated for each row */\
  PetscInt          *ilen;            /* actual length of each row */\
  PetscInt          reallocs;         /* number of mallocs done during MatSetValues() \
                                        as more values are set than were prealloced */\
  PetscInt          rmax;             /* max nonzeros in any row */\
  PetscTruth        keepzeroedrows;   /* keeps matrix structure same in calls to MatZeroRows()*/\
  PetscTruth        ignorezeroentries; \
  PetscInt          *xtoy,*xtoyB;     /* map nonzero pattern of X into Y's, used by MatAXPY() */\
  Mat               XtoY;             /* used by MatAXPY() */\
  PetscTruth        free_ij;          /* free the column indices j and row offsets i when the matrix is destroyed */ \
  PetscTruth        free_a;           /* free the numerical values when matrix is destroy */ \
  Mat_CompressedRow compressedrow;    /* use compressed row format */                      \
  PetscInt          nz;               /* nonzeros */                                       \
  PetscInt          *i;               /* pointer to beginning of each row */               \
  PetscInt          *j;               /* column values: j + i[k] - 1 is start of row k */  \
  PetscInt          *diag;            /* pointers to diagonal elements */                  \
  datatype          *a;               /* nonzero elements */                               \
  PetscScalar       *solve_work;      /* work space used in MatSolve */                    \
  IS                row, col, icol    /* index sets, used for reorderings */               

/*  
  MATSEQAIJ format - Compressed row storage (also called Yale sparse matrix
  format) or compressed sparse row (CSR).  The i[] and j[] arrays start at 0. For example,
  j[i[k]+p] is the pth column in row k.  Note that the diagonal
  matrix elements are stored with the rest of the nonzeros (not separately).
*/

/* Info about i-nodes (identical nodes) helper class for SeqAIJ */
typedef struct {
  MatScalar   *bdiag,*ibdiag;               /* diagonal blocks of matrix used for MatRelax_Inode() */
  PetscInt    bdiagsize;                       /* length of bdiag and ibdiag */
  PetscTruth  ibdiagvalid;                     /* do ibdiag[] and bdiag[] contain the most recent values */

  PetscTruth use;
  PetscInt   node_count;                    /* number of inodes */
  PetscInt   *size;                         /* size of each inode */
  PetscInt   limit;                         /* inode limit */
  PetscInt   max_limit;                     /* maximum supported inode limit */
  PetscTruth checked;                       /* if inodes have been checked for */
} Mat_Inode;

EXTERN PetscErrorCode MatView_Inode(Mat,PetscViewer);
EXTERN PetscErrorCode MatAssemblyEnd_Inode(Mat,MatAssemblyType);
EXTERN PetscErrorCode MatDestroy_Inode(Mat);
EXTERN PetscErrorCode MatCreate_Inode(Mat);
EXTERN PetscErrorCode MatSetOption_Inode(Mat,MatOption,PetscTruth);
EXTERN PetscErrorCode MatDuplicate_Inode(Mat,MatDuplicateOption,Mat*);
EXTERN PetscErrorCode MatILUDTFactor_Inode(Mat,IS,IS,const MatFactorInfo*,Mat*);
EXTERN PetscErrorCode MatLUFactorSymbolic_Inode(Mat,Mat,IS,IS,const MatFactorInfo*);
EXTERN PetscErrorCode MatILUFactorSymbolic_Inode(Mat,Mat,IS,IS,const MatFactorInfo*);


typedef struct {
  SEQAIJHEADER(MatScalar);
  Mat_Inode    inode;
  MatScalar    *saved_values;             /* location for stashing nonzero values of matrix */

  PetscScalar  *idiag,*mdiag,*ssor_work;  /* inverse of diagonal entries, diagonal values and workspace for Eisenstat trick */
  PetscTruth   idiagvalid;                     /* current idiag[] and mdiag[] are valid */
  PetscScalar  fshift,omega;                   /* last used omega and fshift */

  ISColoring   coloring;                  /* set with MatADSetColoring() used by MatADSetValues() */
} Mat_SeqAIJ;

/*
    Frees the a, i, and j arrays from the XAIJ (AIJ, BAIJ, and SBAIJ) matrix types
*/
#undef __FUNCT__  
#define __FUNCT__ "MatSeqXAIJFreeAIJ"
PETSC_STATIC_INLINE PetscErrorCode MatSeqXAIJFreeAIJ(Mat AA,MatScalar **a,PetscInt **j,PetscInt **i) 
{
                                     PetscErrorCode ierr;
                                     Mat_SeqAIJ     *A = (Mat_SeqAIJ*) AA->data;
                                     if (A->singlemalloc) {
                                       ierr = PetscFree3(*a,*j,*i);CHKERRQ(ierr);
                                     } else {
                                       if (A->free_a  && *a) {ierr = PetscFree(*a);CHKERRQ(ierr);}
                                       if (A->free_ij && *j) {ierr = PetscFree(*j);CHKERRQ(ierr);}
                                       if (A->free_ij && *i) {ierr = PetscFree(*i);CHKERRQ(ierr);}
                                     }
                                     *a = 0; *j = 0; *i = 0;
                                     return 0;
                                   }

/*
    Allocates larger a, i, and j arrays for the XAIJ (AIJ, BAIJ, and SBAIJ) matrix types
*/
#define MatSeqXAIJReallocateAIJ(Amat,AM,BS2,NROW,ROW,COL,RMAX,AA,AI,AJ,RP,AP,AIMAX,NONEW,datatype) \
  if (NROW >= RMAX) {\
	Mat_SeqAIJ *Ain = (Mat_SeqAIJ*)Amat->data;\
        /* there is no extra room in row, therefore enlarge */ \
        PetscInt   new_nz = AI[AM] + CHUNKSIZE,len,*new_i=0,*new_j=0; \
        datatype   *new_a; \
 \
        if (NONEW == -2) SETERRQ2(PETSC_ERR_ARG_OUTOFRANGE,"New nonzero at (%D,%D) caused a malloc",ROW,COL); \
        /* malloc new storage space */ \
        ierr = PetscMalloc3(BS2*new_nz,datatype,&new_a,new_nz,PetscInt,&new_j,AM+1,PetscInt,&new_i);CHKERRQ(ierr);\
 \
        /* copy over old data into new slots */ \
        for (ii=0; ii<ROW+1; ii++) {new_i[ii] = AI[ii];} \
        for (ii=ROW+1; ii<AM+1; ii++) {new_i[ii] = AI[ii]+CHUNKSIZE;} \
        ierr = PetscMemcpy(new_j,AJ,(AI[ROW]+NROW)*sizeof(PetscInt));CHKERRQ(ierr); \
        len = (new_nz - CHUNKSIZE - AI[ROW] - NROW); \
        ierr = PetscMemcpy(new_j+AI[ROW]+NROW+CHUNKSIZE,AJ+AI[ROW]+NROW,len*sizeof(PetscInt));CHKERRQ(ierr); \
        ierr = PetscMemcpy(new_a,AA,BS2*(AI[ROW]+NROW)*sizeof(datatype));CHKERRQ(ierr); \
        ierr = PetscMemzero(new_a+BS2*(AI[ROW]+NROW),BS2*CHUNKSIZE*sizeof(datatype));CHKERRQ(ierr);\
        ierr = PetscMemcpy(new_a+BS2*(AI[ROW]+NROW+CHUNKSIZE),AA+BS2*(AI[ROW]+NROW),BS2*len*sizeof(datatype));CHKERRQ(ierr);  \
        /* free up old matrix storage */ \
        ierr = MatSeqXAIJFreeAIJ(A,&Ain->a,&Ain->j,&Ain->i);CHKERRQ(ierr);\
        AA = new_a; \
        Ain->a = (MatScalar*) new_a;		   \
        AI = Ain->i = new_i; AJ = Ain->j = new_j;  \
        Ain->singlemalloc = PETSC_TRUE; \
 \
        RP          = AJ + AI[ROW]; AP = AA + BS2*AI[ROW]; \
        RMAX        = AIMAX[ROW] = AIMAX[ROW] + CHUNKSIZE; \
        Ain->maxnz += CHUNKSIZE; \
        Ain->reallocs++; \
      } \

EXTERN_C_BEGIN
EXTERN PetscErrorCode MatSeqAIJSetPreallocation_SeqAIJ(Mat,PetscInt,PetscInt*);
EXTERN_C_END
EXTERN PetscErrorCode MatILUFactorSymbolic_SeqAIJ(Mat,Mat,IS,IS,const MatFactorInfo*);
EXTERN PetscErrorCode MatICCFactorSymbolic_SeqAIJ(Mat,Mat,IS,const MatFactorInfo*);
EXTERN PetscErrorCode MatCholeskyFactorSymbolic_SeqAIJ(Mat,Mat,IS,const MatFactorInfo*);
EXTERN PetscErrorCode MatCholeskyFactorNumeric_SeqAIJ(Mat,Mat,const MatFactorInfo*);
EXTERN PetscErrorCode MatDuplicate_SeqAIJ(Mat,MatDuplicateOption,Mat*);
EXTERN PetscErrorCode MatMissingDiagonal_SeqAIJ(Mat,PetscTruth*,PetscInt*);
EXTERN PetscErrorCode MatMarkDiagonal_SeqAIJ(Mat);

EXTERN PetscErrorCode MatMult_SeqAIJ(Mat A,Vec,Vec);
EXTERN PetscErrorCode MatMultAdd_SeqAIJ(Mat A,Vec,Vec,Vec);
EXTERN PetscErrorCode MatMultTranspose_SeqAIJ(Mat A,Vec,Vec);
EXTERN PetscErrorCode MatMultTransposeAdd_SeqAIJ(Mat A,Vec,Vec,Vec);
EXTERN PetscErrorCode MatRelax_SeqAIJ(Mat,Vec,PetscReal,MatSORType,PetscReal,PetscInt,PetscInt,Vec);

EXTERN PetscErrorCode MatSetColoring_SeqAIJ(Mat,ISColoring);
EXTERN PetscErrorCode MatSetValuesAdic_SeqAIJ(Mat,void*);
EXTERN PetscErrorCode MatSetValuesAdifor_SeqAIJ(Mat,PetscInt,void*);

EXTERN PetscErrorCode MatGetSymbolicTranspose_SeqAIJ(Mat,PetscInt *[],PetscInt *[]);
EXTERN PetscErrorCode MatGetSymbolicTransposeReduced_SeqAIJ(Mat,PetscInt,PetscInt,PetscInt *[],PetscInt *[]);
EXTERN PetscErrorCode MatRestoreSymbolicTranspose_SeqAIJ(Mat,PetscInt *[],PetscInt *[]);
EXTERN PetscErrorCode MatToSymmetricIJ_SeqAIJ(PetscInt,PetscInt*,PetscInt*,PetscInt,PetscInt,PetscInt**,PetscInt**);
EXTERN PetscErrorCode MatLUFactorSymbolic_SeqAIJ(Mat,Mat,IS,IS,const MatFactorInfo*);
EXTERN PetscErrorCode MatLUFactorNumeric_SeqAIJ(Mat,Mat,const MatFactorInfo*);
EXTERN PetscErrorCode MatLUFactorNumeric_SeqAIJ_InplaceWithPerm(Mat,Mat,const MatFactorInfo*);
EXTERN PetscErrorCode MatLUFactor_SeqAIJ(Mat,IS,IS,const MatFactorInfo*);
EXTERN PetscErrorCode MatSolve_SeqAIJ(Mat,Vec,Vec);
EXTERN PetscErrorCode MatSolve_SeqAIJ_NaturalOrdering(Mat,Vec,Vec);
EXTERN PetscErrorCode MatSolve_SeqAIJ_InplaceWithPerm(Mat,Vec,Vec);
EXTERN PetscErrorCode MatSolveAdd_SeqAIJ(Mat,Vec,Vec,Vec);
EXTERN PetscErrorCode MatSolveTranspose_SeqAIJ(Mat,Vec,Vec);
EXTERN PetscErrorCode MatSolveTransposeAdd_SeqAIJ(Mat,Vec,Vec,Vec);
EXTERN PetscErrorCode MatMatSolve_SeqAIJ(Mat,Mat,Mat);
EXTERN PetscErrorCode MatEqual_SeqAIJ(Mat A,Mat B,PetscTruth* flg);
EXTERN PetscErrorCode MatFDColoringCreate_SeqAIJ(Mat,ISColoring,MatFDColoring);
EXTERN PetscErrorCode MatILUDTFactor_SeqAIJ(Mat,IS,IS,const MatFactorInfo*,Mat*);
EXTERN PetscErrorCode MatLoad_SeqAIJ(PetscViewer, const MatType,Mat*);
EXTERN PetscErrorCode RegisterApplyPtAPRoutines_Private(Mat);
EXTERN PetscErrorCode MatMatMult_SeqAIJ_SeqAIJ(Mat,Mat,MatReuse,PetscReal,Mat*);
EXTERN PetscErrorCode MatMatMultSymbolic_SeqAIJ_SeqAIJ(Mat,Mat,PetscReal,Mat*);
EXTERN PetscErrorCode MatMatMultNumeric_SeqAIJ_SeqAIJ(Mat,Mat,Mat);
EXTERN PetscErrorCode MatPtAPSymbolic_SeqAIJ(Mat,Mat,PetscReal,Mat*);
EXTERN PetscErrorCode MatPtAPNumeric_SeqAIJ(Mat,Mat,Mat);
EXTERN PetscErrorCode MatPtAPSymbolic_SeqAIJ_SeqAIJ(Mat,Mat,PetscReal,Mat*);
EXTERN PetscErrorCode MatPtAPNumeric_SeqAIJ_SeqAIJ(Mat,Mat,Mat);
EXTERN PetscErrorCode MatMatMultTranspose_SeqAIJ_SeqAIJ(Mat,Mat,MatReuse,PetscReal,Mat*);
EXTERN PetscErrorCode MatMatMultTransposeSymbolic_SeqAIJ_SeqAIJ(Mat,Mat,PetscReal,Mat*);
EXTERN PetscErrorCode MatMatMultTransposeNumeric_SeqAIJ_SeqAIJ(Mat,Mat,Mat);
EXTERN PetscErrorCode MatSetValues_SeqAIJ(Mat,PetscInt,const PetscInt[],PetscInt,const PetscInt[],const PetscScalar[],InsertMode);
EXTERN PetscErrorCode MatGetRow_SeqAIJ(Mat,PetscInt,PetscInt*,PetscInt**,PetscScalar**);
EXTERN PetscErrorCode MatRestoreRow_SeqAIJ(Mat,PetscInt,PetscInt*,PetscInt**,PetscScalar**);
EXTERN PetscErrorCode MatAXPY_SeqAIJ(Mat,PetscScalar,Mat,MatStructure);
EXTERN PetscErrorCode MatGetRowIJ_SeqAIJ(Mat,PetscInt,PetscTruth,PetscTruth,PetscInt*,PetscInt *[],PetscInt *[],PetscTruth *);
EXTERN PetscErrorCode MatRestoreRowIJ_SeqAIJ(Mat,PetscInt,PetscTruth,PetscTruth,PetscInt *,PetscInt *[],PetscInt *[],PetscTruth *);
EXTERN PetscErrorCode MatGetColumnIJ_SeqAIJ(Mat,PetscInt,PetscTruth,PetscTruth,PetscInt*,PetscInt *[],PetscInt *[],PetscTruth *);
EXTERN PetscErrorCode MatRestoreColumnIJ_SeqAIJ(Mat,PetscInt,PetscTruth,PetscTruth,PetscInt *,PetscInt *[],PetscInt *[],PetscTruth *);
EXTERN PetscErrorCode MatDestroy_SeqAIJ(Mat);
EXTERN PetscErrorCode MatView_SeqAIJ(Mat,PetscViewer);

EXTERN_C_BEGIN
EXTERN PetscErrorCode PETSCMAT_DLLEXPORT MatConvert_SeqAIJ_SeqSBAIJ(Mat,const MatType,MatReuse,Mat*);
EXTERN PetscErrorCode PETSCMAT_DLLEXPORT MatConvert_SeqAIJ_SeqBAIJ(Mat,const MatType,MatReuse,Mat*);
EXTERN PetscErrorCode PETSCMAT_DLLEXPORT MatConvert_SeqAIJ_SeqCSRPERM(Mat,const MatType,MatReuse,Mat*);
EXTERN PetscErrorCode PETSCMAT_DLLEXPORT MatReorderForNonzeroDiagonal_SeqAIJ(Mat,PetscReal,IS,IS);
EXTERN_C_END

#endif
