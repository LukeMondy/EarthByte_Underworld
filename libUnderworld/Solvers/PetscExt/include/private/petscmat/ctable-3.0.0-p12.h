/* Contributed by - Mark Adams */

#if !defined(__CTABLE300_H)
#define __CTABLE300_H

struct _n_PetscTable {
  PetscInt *keytable;
  PetscInt *table;
  PetscInt count;
  PetscInt tablesize;
  PetscInt head;
};

#include "petscctable.h"
#endif
