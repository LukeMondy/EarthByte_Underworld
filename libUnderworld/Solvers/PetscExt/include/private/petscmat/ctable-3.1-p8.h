/* Contributed by - Mark Adams */

#if !defined(__CTABLE31_H)
#define __CTABLE31_H

struct _n_PetscTable {
  PetscInt *keytable;
  PetscInt *table;
  PetscInt count;
  PetscInt tablesize;
  PetscInt head;
};

#include "petscctable.h"
#endif
