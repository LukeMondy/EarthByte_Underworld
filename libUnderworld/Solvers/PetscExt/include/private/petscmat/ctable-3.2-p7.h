/* Contributed by - Mark Adams */

#if !defined(__CTABLE_H)
#define __CTABLE_H

struct _n_PetscTable {
  PetscInt *keytable;
  PetscInt *table;
  PetscInt count;
  PetscInt tablesize;
  PetscInt head;
};

#include <petscctable.h>
#endif
