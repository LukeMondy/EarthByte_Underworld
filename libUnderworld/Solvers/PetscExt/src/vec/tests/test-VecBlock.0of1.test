EXEC 		= ./test-VecBlock.app
NPROCS 		= 1
TEST_TYPE       = tolerance_diff_pl,1.0e-10
