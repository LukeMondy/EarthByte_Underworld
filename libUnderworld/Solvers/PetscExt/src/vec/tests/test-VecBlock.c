/*@ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 **
 **    Copyright (c) 2007 - 2008, 
 **        Dave A. May [dave.mayhem23@gmail.com]
 **        School of Mathematical Sciences, 
 **        Monash University
 **        Victoria, 3800
 **        Australia
 **
 **    Copyright (c) 2009 - 2010, 
 **        Dave A. May [dave.mayhem23@gmail.com]
 **        Geophysical Fluid Dynamics, 
 **        Department of Earth Sciences,
 **        ETH ZÃ¼rich,
 **        Switzerland
 **
 **    Project:       PetscExt_v3.1
 **    Description:   A collection of extensions to the PETSc library.
 **
 **    All rights reserved.
 **    Redistribution and use in source and binary forms, with or without modification,
 **    are permitted provided that the following conditions are met:
 **
 **        * Redistributions of source code must retain the above copyright notice,
 **              this list of conditions and the following disclaimer.
 **        * Redistributions in binary form must reproduce the above copyright
 **              notice, this list of conditions and the following disclaimer in the
 **              documentation and/or other materials provided with the distribution.
 **        * Neither the name of the Monash University nor the names of its contributors
 **              may be used to endorse or promote products derived from this software
 **              without specific prior written permission.
 **
 **    THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
 **    "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO,
 **    THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
 **    PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS
 **    BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
 **    CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
 **    SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 **    HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 **    LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT 
 **    OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 **
 **
 **    $Id$
 **
 ** ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~@*/


#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <math.h>
#include <stdarg.h>

#include <petsc.h>
#include <petscvec.h>
#include <petscmat.h>
#include <petscksp.h>

#include "petscext_vec.h"



/* 
Usage;

VecCreate( );
VecSetSizes( PETSC_DECIDE, 2 ); // block size 
VecSetType( v, "block" ); // calls VecCreate_Block


*/


/*
X = ( a b )^T
a = ( c d )^T
b = ( e f )^T

c,d,e,f = seq


*/

void test_view( void )
{
	Vec X, a,b;
	Vec c,d,e,f;
	PetscInt index;
	PetscReal val;
	
	
	PetscPrintf( PETSC_COMM_WORLD, "\n\n============== %s ==============\n", __func__ );
	
	
	VecCreate( PETSC_COMM_WORLD, &X );
	VecSetSizes( X, 2, 2 );
	VecSetType( X, "block" );
	
	VecCreate( PETSC_COMM_WORLD, &a );
	VecSetSizes( a, 2, 2 );
	VecSetType( a, "block" );
	
	VecCreate( PETSC_COMM_WORLD, &b );
	VecSetSizes( b, 2, 2 );
	VecSetType( b, "block" );
	
	/* assemble X */
	VecBlockSetValue( X, 0, a, INSERT_VALUES );
	VecBlockSetValue( X, 1, b, INSERT_VALUES );
	VecAssemblyBegin(X);
	VecAssemblyEnd(X);
	
	
	VecCreate( PETSC_COMM_WORLD, &c );
	VecSetSizes( c, 3, 3 );
	VecSetType( c, VECSEQ );
	VecDuplicate( c, &d );
	VecDuplicate( c, &e );
	VecDuplicate( c, &f );
	
	VecSet( c, 1.0 );
	VecSet( d, 2.0 );
	VecSet( e, 3.0 );
	VecSetRandom( f, PETSC_NULL );
	VecScale( f, 10.0 );
	
	/* assemble a */
	VecBlockSetValue( a, 0, c, INSERT_VALUES );
	VecBlockSetValue( a, 1, d, INSERT_VALUES );
	VecAssemblyBegin(a);
	VecAssemblyEnd(a);
	
	
	/* assemble b */
	VecBlockSetValue( b, 0, e, INSERT_VALUES );
	VecBlockSetValue( b, 1, f, INSERT_VALUES );
	VecAssemblyBegin(b);
	VecAssemblyEnd(b);
	
//	PetscPrintf( PETSC_COMM_WORLD, "X \n");
//	VecView( X, PETSC_VIEWER_STDOUT_WORLD );
	
//	PetscPrintf( PETSC_COMM_WORLD, "b \n");
//	VecView( b, PETSC_VIEWER_STDOUT_WORLD );
	
	
	VecMax( b, &index, &val );
	PetscPrintf( PETSC_COMM_WORLD, "(max-b) = %f : index = %d \n", val, index );
	
	VecMin( b, &index, &val );
	PetscPrintf( PETSC_COMM_WORLD, "(min-b) = %f : index = %d \n", val, index );
	
	VecMax( X, &index, &val );
	PetscPrintf( PETSC_COMM_WORLD, "(max-X) = %f : index = %d \n", val, index );
	VecMin( X, &index, &val );
	PetscPrintf( PETSC_COMM_WORLD, "(min-X) = %f : index = %d \n", val, index );
	
	PetscViewerSetFormat( PETSC_VIEWER_STDOUT_WORLD, PETSC_VIEWER_ASCII_MATLAB_INFO_DETAIL );
	VecView( X, PETSC_VIEWER_STDOUT_WORLD );
}


void test_vec_ops( void )
{
	Vec X, a,b;
	Vec c,d,e,f;
	PetscInt index;
	PetscScalar val;
	
	
	PetscPrintf( PETSC_COMM_WORLD, "\n\n============== %s ==============\n", __func__ );
	
	VecCreate( PETSC_COMM_WORLD, &X );
	VecSetSizes( X, 2, 2 );
	VecSetType( X, "block" );
	
	VecCreate( PETSC_COMM_WORLD, &a );
	VecSetSizes( a, 2, 2 );
	VecSetType( a, "block" );
	
	VecCreate( PETSC_COMM_WORLD, &b );
	VecSetSizes( b, 2, 2 );
	VecSetType( b, "block" );
	
	/* assemble X */
	VecBlockSetValue( X, 0, a, INSERT_VALUES );
	VecBlockSetValue( X, 1, b, INSERT_VALUES );
	VecAssemblyBegin(X);
	VecAssemblyEnd(X);
	
	
	
	VecCreate( PETSC_COMM_WORLD, &c );
	VecSetSizes( c, 3, 3 );
	VecSetType( c, VECSEQ );
	VecDuplicate( c, &d );
	VecDuplicate( c, &e );
	VecDuplicate( c, &f );
	
	VecSet( c, 1.0 );
	VecSet( d, 2.0 );
	VecSet( e, 3.0 );
	VecSet( f, 4.0 );
	
	/* assemble a */
	VecBlockSetValue( a, 0, c, INSERT_VALUES );
	VecBlockSetValue( a, 1, d, INSERT_VALUES );
	VecAssemblyBegin(a);
	VecAssemblyEnd(a);
	
	/* assemble b */
	VecBlockSetValue( b, 0, e, INSERT_VALUES );
	VecBlockSetValue( b, 1, f, INSERT_VALUES );
	VecAssemblyBegin(b);
	VecAssemblyEnd(b);
	
//	PetscPrintf( PETSC_COMM_WORLD, "X \n");
//	VecView( X, PETSC_VIEWER_STDOUT_WORLD );
	
	VecDot( X,X, &val );
	PetscPrintf( PETSC_COMM_WORLD, "X.X = %f \n", val ); 
	
}



Vec gen_test_vector( MPI_Comm comm, PetscInt length, PetscInt start_value, PetscInt stride )
{
	int nproc;
	Vec v;
	PetscInt i;
	PetscScalar vx;
	
	MPI_Comm_size( comm, &nproc );
	
	VecCreate( comm, &v );
	VecSetSizes( v, PETSC_DECIDE, length );
	if( nproc == 1 ) VecSetType( v, VECSEQ );
	else VecSetType( v, VECMPI );
	
	for( i=0; i<length; i++ ) {
		vx = (PetscScalar)( start_value + i * stride );
		VecSetValue( v, i, vx, INSERT_VALUES );
	}
	VecAssemblyBegin( v );
	VecAssemblyEnd( v );
	
	return v;
	
}


/*
X = ( [0,1,2,3], [10,12,14,16,18] )
Y = ( [4,7,10,13], [5,6,7,8,9] )

Y = aX + y = ( [4,8,12,16], (15,18,21,24,27] )
Y = aX + y = ( [4,9,14,19], (25,30,35,40,45] )

*/

void test_axpy_dot_max( void )
{
	Vec x1,y1, x2,y2;
	Vec X, Y;
	PetscReal real;
	PetscScalar scalar;
	PetscInt index;
	
	
	x1 = gen_test_vector( PETSC_COMM_WORLD, 4, 0, 1 );
	x2 = gen_test_vector( PETSC_COMM_WORLD, 5, 10, 2 );
	
	y1 = gen_test_vector( PETSC_COMM_WORLD, 4, 4, 3 );
	y2 = gen_test_vector( PETSC_COMM_WORLD, 5, 5, 1 );
	
	
	
	VecCreate( PETSC_COMM_WORLD, &X );
	VecSetSizes( X, 2,2 );
	VecSetType( X, "block" );
	VecBlockSetValue( X, 0, x1, INSERT_VALUES );
	VecBlockSetValue( X, 1, x2, INSERT_VALUES );
	VecAssemblyBegin(X);
	VecAssemblyEnd(X);
	Stg_VecDestroy( & x1 );			Stg_VecDestroy( & x2 );
	
	
	VecCreate( PETSC_COMM_WORLD, &Y );
	VecSetSizes( Y, 2,2 );
	VecSetType( Y, "block" );
	VecBlockSetValue( Y, 0, y1, INSERT_VALUES );
	VecBlockSetValue( Y, 1, y2, INSERT_VALUES );
	VecAssemblyBegin(Y);
	VecAssemblyEnd(Y);
	Stg_VecDestroy( & y1 );			Stg_VecDestroy( & y2 );
	
	
	PetscPrintf( PETSC_COMM_WORLD, "VecAXPY \n");
	VecAXPY( Y, 1.0, X ); /* Y <- a X + Y */
	VecBlockGetSubVector( Y, 0, &y1 );					VecBlockGetSubVector( Y, 1, &y2 );
	PetscPrintf( PETSC_COMM_WORLD, "(1) y1 = \n" );		VecView( y1, PETSC_VIEWER_STDOUT_WORLD );
	PetscPrintf( PETSC_COMM_WORLD, "(1) y2 = \n" );		VecView( y2, PETSC_VIEWER_STDOUT_WORLD );
	VecBlockRestoreSubVectors( Y );
	VecDot( X,Y, &scalar );
#ifdef PETSC_USE_COMPLEX
	PetscPrintf( PETSC_COMM_WORLD, "X.Y = %lf + %lfi \n", PetscRealPart(scalar), PetscImaginaryPart(scalar) ); 
#else
	PetscPrintf( PETSC_COMM_WORLD, "X.Y = %lf \n", scalar ); 
#endif 
	
	
	VecAXPY( Y, 1.0, X ); /* Y <- a X + Y */
	VecBlockGetSubVector( Y, 0, &y1 );					VecBlockGetSubVector( Y, 1, &y2 );
	PetscPrintf( PETSC_COMM_WORLD, "(2) y1 = \n" );		VecView( y1, PETSC_VIEWER_STDOUT_WORLD );
	PetscPrintf( PETSC_COMM_WORLD, "(2) y2 = \n" );		VecView( y2, PETSC_VIEWER_STDOUT_WORLD );
	VecBlockRestoreSubVectors( Y );
	VecDot( X,Y, &scalar );
#ifdef PETSC_USE_COMPLEX
	PetscPrintf( PETSC_COMM_WORLD, "X.Y = %lf + %lfi \n", PetscRealPart(scalar), PetscImaginaryPart(scalar) ); 
#else
	PetscPrintf( PETSC_COMM_WORLD, "X.Y = %lf \n", scalar ); 
#endif 
	
	
	
	VecMax( X, &index, &real );
	PetscPrintf( PETSC_COMM_WORLD, "(max-X) = %f : index = %d \n", real, index );
	VecMin( X, &index, &real );
	PetscPrintf( PETSC_COMM_WORLD, "(min-X) = %f : index = %d \n", real, index );
	
//	PetscViewerSetFormat( PETSC_VIEWER_STDOUT_WORLD, PETSC_VIEWER_ASCII_MATLAB_INFO_DETAIL );
//	VecView( X, PETSC_VIEWER_STDOUT_WORLD );
	
	
	Stg_VecDestroy( & X );
	Stg_VecDestroy( & Y );
	
}



int main( int argc, char **args )
{
	int i;
	int BIG=100000;
	
	PetscInitialize( &argc, &args,(char *)0, PETSC_NULL );
	PetscExtVecRegisterAll();
	
//	test_view();	// this test will output pointers, thus we cannot diff result
	test_axpy_dot_max();
	
	PetscExtVecRegisterDestroyAll();
	PetscFinalize();
	return 0;
}
