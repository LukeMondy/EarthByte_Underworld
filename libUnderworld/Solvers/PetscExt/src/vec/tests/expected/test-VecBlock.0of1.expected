VecAXPY 
(1) y1 = 
4
8
12
16
(1) y2 = 
15
18
21
24
27
X.Y = 1610.000000 
(2) y1 = 
4
9
14
19
(2) y2 = 
25
30
35
40
45
X.Y = 2644.000000 
(max-X) = 18.000000 : index = 8 
(min-X) = 0.000000 : index = 0 
